server/tests/trim_gap_handling.rs

```diff
@@ -0,0 +1,354 @@
+        base_config.clone(),
+        BinarySource::CargoTest,
+        enum_set!(Role::HttpIngress | Role::Worker),
+    );
+    *worker_3.metadata_store_client_mut() = MetadataStoreClient::Embedded {
+        address: cluster.nodes[0].node_address().clone(),
+    };
+
+    let ingress_url = format!(
+        "http://{}/Counter/0/get",
+        worker_3.config().ingress.bind_address
+    );
+    let mut worker_3_imported_snapshot = worker_3.lines(
+        format!(
+            "Importing partition store snapshot.*{}",
+            snapshot_response.snapshot_id
+        )
+        .parse()?,
+    );
+    let mut worker_3 = worker_3
+        .start_clustered(cluster.base_dir(), cluster.cluster_name())
+        .await?;
+    assert!(
+        tokio::time::timeout(Duration::from_secs(20), worker_3_imported_snapshot.next())
+            .await
+            .is_ok()
+    );
+
+    // todo(pavel): promote node 3 to be the leader for partition 0 and invoke the service again
+    // right now, all we are asserting is that the new node is applying newly appended log records
+    assert!(http_client
+        .post(ingress_url)
+        .send()
+        .await?
+        .status()
+        .is_success());
+    applied_lsn_converged(&mut client, Duration::from_secs(5), 3, 0).await?;
+
+    worker_3.graceful_shutdown(Duration::from_secs(1)).await?;
+    cluster.graceful_shutdown(Duration::from_secs(1)).await?;
+    Ok(())
+}
+
+async fn any_partition_active(
+    client: &mut ClusterCtrlSvcClient<Channel>,
+    timeout: Duration,
+) -> googletest::Result<()> {
+    let deadline = tokio::time::Instant::now() + timeout;
+    loop {
+        let cluster_state = client
+            .get_cluster_state(ClusterStateRequest {})
+            .await?
+            .into_inner()
+            .cluster_state
+            .unwrap();
+
+        if cluster_state.nodes.values().any(|n| {
+            n.state.as_ref().is_some_and(|s| match s {
+                State::Alive(s) => s.partitions.values().any(|p| {
+                    RunMode::try_from(p.effective_mode).is_ok_and(|m| m == RunMode::Leader)
+                }),
+                _ => false,
+            })
+        }) {
+            break;
+        }
+        if tokio::time::Instant::now() > deadline {
+            fail!(
+                "Partition processor did not become ready within {:?}",
+                timeout
+            )?;
+        }
+        tokio::time::sleep(Duration::from_millis(250)).await;
+    }
+    Ok(())
+}
+
+async fn trim_log(
+    client: &mut ClusterCtrlSvcClient<Channel>,
+    trim_point: u64,
+    timeout: Duration,
+) -> googletest::Result<()> {
+    let deadline = tokio::time::Instant::now() + timeout;
+    loop {
+        let response = client
+            .trim_log(TrimLogRequest {
+                log_id: 0,
+                trim_point,
+            })
+            .await?
+            .into_inner();
+
+        if response.trim_point.is_some_and(|tp| tp == trim_point) {
```

Comment:
Maybe this hardens this function a bit?

```suggestion
        if response.trim_point.is_some_and(|tp| tp >= trim_point) {
```

Reply:
Nice!